// nativescript-socket.io/socket.ios.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("./common");
exports.enableDebug = common_1.enableDebug;
exports.disableDebug = common_1.disableDebug;
const helpers_1 = require("./helpers");
SocketIOClient;
SocketAckEmitter;
OnAckCallback;
SocketEngine;
const NAMESPACE_REGEXP = /^https?\:\/\/[^\/]*(\/.*)$/i;
class Socket extends common_1.SocketBase {
    constructor(uri, options = {}) {
        super();
        let _options = {};
        if (NAMESPACE_REGEXP.test(uri)) {
            _options.nsp = uri.match(NAMESPACE_REGEXP)[1] || '/';
        }
        else {
            _options.nsp = '/';
        }
        if (options.query) {
            _options.connectParams = {};
            if (typeof options.query === 'string') {
                options.query.split('&').forEach(function (pair) {
                    pair = pair.split('=').map(decodeURIComponent);
                    _options.connectParams[pair[0]] = pair[1];
                });
            }
            else {
                Object.keys(options.query).forEach(function (key) {
                    _options.connectParams[key] = String(options.query[key]);
                });
            }
        }
        if (options.ios) {
            Object.keys(options.ios).forEach(function (prop) {
                _options[prop] = options.ios[prop];
            });
        }
        this.ios = SocketIOClient.alloc().initWithSocketURLConfig(NSURL.URLWithString(uri), _options);
    }
    get connected() {
        return this.ios && this.ios.engine.connected;
    }
    connect() {
        this.ios.connect();
    }
    disconnect() {
        this.ios.disconnect();
    }
    on(event, callback) {
        let listener = function (data, ack) {
            let payload = helpers_1.deserialize(data);
            common_1.debug('on', event, payload, ack && ack.expected ? 'ack' : '');
            if (ack && ack.expected) {
                let _ack = function (...args) {
                    common_1.debug('on', event, 'ack', args);
                    args = args.map(helpers_1.serialize);
                    ack.with(args);
                };
                payload.push(_ack);
            }
            callback.apply(null, payload);
        };
        let listenerId = this.ios.onCallback(event, listener);
        this._listeners.set(callback, listenerId);
        return this;
    }
    once(event, callback) {
        let listener = function (data, ack) {
            let payload = helpers_1.deserialize(data);
            common_1.debug('once', event, payload, ack && ack.expected ? 'ack' : '');
            if (ack && ack.expected) {
                let _ack = function (...args) {
                    common_1.debug('once', event, 'ack', args);
                    args = args.map(helpers_1.serialize);
                    ack.with(args);
                };
                payload.push(_ack);
            }
            callback.apply(null, payload);
        };
        let listenerId = this.ios.onceCallback(event, listener);
        this._listeners.set(callback, listenerId);
        return this;
    }
    off(event, callback) {
        common_1.debug('off', event, callback);
        if (callback) {
            let listenerId = this._listeners.get(callback);
            if (listenerId) {
                this.ios.offWithId(listenerId);
                this._listeners.delete(callback);
            }
        }
        else {
            this.ios.off(event);
        }
        return this;
    }
    emit(event, ...payload) {
        let ack = payload.pop();
        if (typeof ack === 'undefined') {
            ack = null;
        }
        else if (typeof ack !== 'function') {
            payload.push(ack);
            ack = null;
        }
        common_1.debug('emit', event, payload, ack ? 'ack' : '');
        payload = payload.map(helpers_1.serialize);
        if (ack) {
            let _ack = function (args) {
                args = helpers_1.deserialize(args);
                common_1.debug('emit', event, 'ack', args);
                ack.apply(null, args);
            };
            this.ios.emitWithAckWith(event, payload).timingOutAfterCallback(0, _ack);
        }
        else {
            this.ios.emitWith(event, payload);
        }
        return this;
    }
    removeAllListeners() {
        this.ios.removeAllHandlers();
        return this;
    }
}
exports.Socket = Socket;
function connect(uri, options) {
    let socket = new Socket(uri, options || {});
    socket.connect();
    return socket;
}
exports.connect = connect;
